src/main/java/appeng/recipes/factories/recipes/CuttingRecipeFactory.java

```diff
@@ -0,0 +1,162 @@
+		}
+
+		if( ings.isEmpty() )
+		{
+			throw new JsonParseException( "No ingredients for shapeless recipe" );
+		}
+
+		JsonObject resultObject = (JsonObject) json.get( "result" );
+		int count = JsonUtils.getInt( resultObject, "count", 1 );
+
+		ItemStack itemStack = ItemStack.EMPTY;
+		if( resultObject.has( "part" ) )
+		{
+			String ingredient = resultObject.get( "part" ).getAsString();
+			Object result = (Object) Api.INSTANCE.registries().recipes().resolveItem( AppEng.MOD_ID, ingredient );
+			if( result instanceof ResolverResult )
+			{
+				ResolverResult resolverResult = (ResolverResult) result;
+
+				Item item = Item.getByNameOrId( AppEng.MOD_ID + ":" + resolverResult.itemName );
+
+				if( item == null )
+				{
+					AELog.warn( "item was null for " + resolverResult.itemName + " ( " + ingredient + " )!" );
+					throw new JsonSyntaxException( "Got a null item for " + resolverResult.itemName + " ( " + ingredient + " ). This should never happen!" );
+				}
+
+				itemStack = new ItemStack( item, count, resolverResult.damageValue, resolverResult.compound );
+			}
+			else
+			{
+				throw new JsonSyntaxException( "Couldn't find the resulting item in AE. This means AE was provided a recipe that it shouldn't be handling.\n" + "Was looking for : '" + ingredient + "'." );
+			}
+		}
+		else if( resultObject.has( "item" ) )
+		{
+			itemStack = CraftingHelper.getItemStack( JsonUtils.getJsonObject( json, "result" ), context );
+		}
+
+		if( !itemStack.isEmpty() )
+		{
+			return new CuttingRecipe( group.isEmpty() ? null : new ResourceLocation( group ), ings, itemStack );
+		}
+		else
+		{
+			throw new JsonSyntaxException( "Could not parse result" );
+		}
+	}
+
+	private static class CuttingRecipe extends ShapelessOreRecipe
+	{
+		public CuttingRecipe( ResourceLocation group, NonNullList<Ingredient> input, ItemStack result )
+		{
+			super( group, input, result );
+		}
+
+		private boolean isKnife( ItemStack stack )
+		{
+			return Api.INSTANCE.definitions().items().certusQuartzKnife().isSameAs( stack ) || Api.INSTANCE.definitions().items().netherQuartzKnife().isSameAs(
+					stack );
+		}
+
+		@Override
+		public boolean canFit( int width, int height )
+		{
+			return width * height >= this.input.size() + 1;
+		}
+
+		@Override
+		public boolean matches( @Nonnull InventoryCrafting inv, @Nonnull World world )
+		{
+			ItemStack knife = ItemStack.EMPTY;
+			NonNullList<Ingredient> required = NonNullList.create();
+			required.addAll( input );
+
+			for( int i = 0; i < inv.getSizeInventory(); ++i )
+			{
+				final ItemStack current = inv.getStackInSlot( i );
+				if( !current.isEmpty() )
+				{
+					if( this.isKnife( current ) )
+					{
+						if( knife.isEmpty() )
+						{
+							knife = current;
+						}
+						else
+						{
+							// multiple knifes
+							return false;
+						}
+					}
+					else
+					{
+						boolean inRecipe = false;
+						final Iterator<Ingredient> req = required.iterator();
+
+						while( req.hasNext() )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)